File: src/fixpm/detector.py
"""Classify what went wrong in a failed command line.

One generic engine consumes every ``ManagerSpec``; per-CLI knowledge lives
entirely in the rule tables under ``fixpm.rules``. The engine must never see
output of the failed command — unlike tools that re-run it to read the error,
fixpm classifies from the command line alone, which is what keeps it under
~100 ms and free of side effects.
"""

from __future__ import annotations

import re
import shlex

from .distance import rank
from .rules import spec_for_binary
from .rules.base import Issue, IssueKind, ManagerSpec

_PKG_RE = re.compile(
    r"^(@[a-z0-9-~][a-z0-9-._~]*/)?[a-z0-9-~][a-z0-9-._~]*$", re.IGNORECASE
)


def tokenize(text: str) -> list[str]:
    try:
        return shlex.split(text)
    except ValueError:
        return text.split()


def looks_like_package(token: str) -> bool:
    """Conservative npm-package-name shape check (paths, globs and files out)."""
    if not token or any(c in token for c in "*~=\\"):
        return False
    if token.startswith((".", "~")):
        return False
    if "/" in token and not token.startswith("@"):
        return False
    return bool(_PKG_RE.match(token))


def _nearest(token: str, pool: tuple[str, ...], top: int = 2,
             min_score: float = 0.6) -> list[str]:
    return [c for c, _ in rank(token.lower(), [p.lower() for p in sorted(pool)],
                               top=top, min_score=min_score)]


def _hinted(token: str, spec: ManagerSpec, pool: set[str] | tuple[str, ...],
            ) -> str | None:
    """The curated hint for *token*, if it is valid in this slot.

    One ``typo_hints`` map feeds both the command slot and the chain-verb slot,
    so the target is checked against the slot's own vocabulary. Without that,
    a hint meant for one slot leaks into the other and suggests an invalid
    command: ``lss -> ls`` is a docker object verb, not a top-level one.
    """
    hint = spec.typo_hints.get(token.lower())
    return hint if hint is not None and hint in pool else None


def _sub_candidates(token: str, spec: ManagerSpec) -> list[str]:
    hint = _hinted(token, spec, spec.vocabulary)
    if hint is not None:
        return [hint]
    return _nearest(token, tuple(spec.vocabulary), top=3,
                    min_score=spec.subcommand_min_score)


def analyze(
    text: str, spec: ManagerSpec | None = None
) -> tuple[list[str], ManagerSpec | None, list[Issue]]:
    """Return ``(tokens, spec_or_None, issues)`` for a failed command line."""
    tokens = tokenize(text)
    located = _locate(tokens, spec)
    if located is None:
        return tokens, None, []
    index, spec = located
    if index < 0:
        # Forced table whose binary is absent from the line. Synthesise it so
        # the suggestion is a command you can actually run rather than a
        # fragment: --manager git "comit -m x" -> "git commit -m x".
        tokens = [spec.binaries[0], *tokens]
        index = 0
    if spec.package_first:
        return tokens, spec, _analyze_package_first(spec, tokens, index + 1)
    return tokens, spec, _analyze_subcommand(spec, tokens, index + 1)


def _locate(tokens: list[str],
            spec: ManagerSpec | None) -> tuple[int, ManagerSpec] | None:
    if spec is not None:
        for i, t in enumerate(tokens[:4]):
            if t in spec.binaries:
                return i, spec
        # Binary absent: -1 tells analyze() to synthesise it. Returning None
        # here made every bare fragment answer "No fix found", which left
        # --manager usable only when the binary was already in the line (and
        # then auto-detection would have found it anyway).
        return -1, spec
    for i, t in enumerate(tokens[:4]):
        found = spec_for_binary(t)
        if found is not None:
            return i, found
    return None


def _analyze_subcommand(spec: ManagerSpec, tokens: list[str],
                        start: int) -> list[Issue]:
    issues: list[Issue] = []
    rest = tokens[start:]

    # Leading global flags, e.g. `npm --prefix ./x install`
    i = 0
    while i < len(rest) and rest[i].startswith("-"):
        flag = rest[i]
        if flag in spec.value_flags and i + 1 < len(rest):
            i += 2
            continue
        if flag not in spec.global_flags:
            near = _nearest(flag, spec.global_flags)
            if near:
                issues.append(Issue(
                    IssueKind.FLAG_TYPO, flag, start + i,
                    f"Unknown global flag '{flag}' — did you mean {near[0]}?",
                    tuple(near),
                ))
        i += 1

    if i >= len(rest):
        return issues

    sub = rest[i]
    sub_abs = start + i
    canon = spec.canonical(sub)

    if canon is None:
        candidates = _sub_candidates(sub, spec)
        if candidates:
            issues.append(Issue(
                IssueKind.SUBCOMMAND_TYPO, sub, sub_abs,
                f"'{sub}' is not a {spec.name} command",
                tuple(candidates),
            ))
        # One fix at a time: stop validating after an unknown command.
        return issues

    # Chain commands like `yarn global add <pkg>`, `docker container ls` or
    # `go mod tidy` — absorb the second-level verb. Which commands chain is a
    # per-CLI fact (spec.chains), so the engine only does the lookup; the slot
    # gets typo tolerance too, because `yarn global ad x` must still resolve
    # the chain (and report the typo) instead of falling apart.
    extra = 0
    verbs = spec.chains.get(canon, ())
    if verbs and i + 1 < len(rest):
        nxt = rest[i + 1]
        if nxt in verbs:
            extra = 1
            canon = f"{canon} {nxt}"
        else:
            # Short transpositions (`ud` -> `up`, `ad` -> `add`) score 0.667
            # and land just under every similarity floor, so curated hints
            # cover them here too — validated against this slot.
            hint = _hinted(nxt, spec, verbs)
            near = [hint] if hint is not None else _nearest(
                nxt, verbs, top=1, min_score=spec.subcommand_min_score)
            if near:
                issues.append(Issue(
                    IssueKind.SUBCOMMAND_TYPO, nxt, start + i + 1,
                    f"'{nxt}' is not a {spec.name} subcommand of '{canon}'",
                    tuple(near),
                ))
                extra = 1
                canon = f"{canon} {near[0]}"
    effective = canon.rsplit(" ", 1)[-1]
    allowed = spec.flags.get(effective, ())

    positional: list[tuple[int, str]] = []
    j = i + 1 + extra
    while j < len(rest):
        token = rest[j]
        if token == "--":  # passthrough separator: everything after is exempt
            break
        if token.startswith("-"):
            if allowed and token not in allowed and token not in spec.global_flags:
                near = _nearest(token, allowed)
                if near:
                    issues.append(Issue(
                        IssueKind.FLAG_TYPO, token, start + j,
                        f"Unknown flag '{token}' for `{spec.name} "
                        f"{effective}` — did you mean {near[0]}?",
                        tuple(near),
                    ))
            if token in spec.value_flags:
                j += 1  # skip the consumed value
        else:
            positional.append((j, token))
        j += 1

    # `npm install express save-dev` -> missing dashes on `save-dev`
    dashless = {f.lstrip("-"): f for f in allowed}
    flagged: set[int] = set()
    for jj, token in positional:
        if len(token) > 1 and token in dashless:
            issues.append(Issue(
                IssueKind.MISSING_DASHES, token, start + jj,
                f"Missing flag prefix — did you mean {dashless[token]}?",
                (dashless[token],),
            ))
            flagged.add(jj)
    positional = [(jj, t) for jj, t in positional if jj not in flagged]

    # Only report a missing argument when nothing else was flagged. For
    # `npm uninstall save-dev` the real problem is the missing dashes, and
    # pairing it with "needs an argument" emitted a second suggestion that
    # still contained the very typo the first one fixes:
    #   [missing --]       npm uninstall --save-dev
    #   [missing argument] npm uninstall save-dev <package>
    if (canon in spec.arg_required and not positional and not flagged
            and extra == 0):
        issues.append(Issue(
            IssueKind.ARG_REQUIRED, sub, sub_abs,
            f"`{spec.name} {sub}` needs an argument "
            f"(see `{spec.name} help {effective}`)",
            (),
        ))

    # Chained package commands are listed explicitly ("global add"), so this
    # stays a plain membership test — the engine holds no idea of which chain
    # names happen to take packages.
    if canon in spec.package_commands:
        checked = 0
        for jj, token in positional:
            if checked >= 2:
                break
            if token.lower() in spec.vocabulary or not looks_like_package(token):
                continue
            issues.append(Issue(
                IssueKind.PACKAGE_TYPO, token, start + jj,
                f"Check package name '{token}'",
                (),
            ))
            checked += 1
    return issues


def _analyze_package_first(spec: ManagerSpec, tokens: list[str],
                           start: int) -> list[Issue]:
    issues: list[Issue] = []
    rest = tokens[start:]
    i = 0
    while i < len(rest) and rest[i].startswith("-"):
        flag = rest[i]
        if flag in spec.value_flags and i + 1 < len(rest):
            i += 2
            continue
        if flag not in spec.global_flags:
            near = _nearest(flag, spec.global_flags)
            if near:
                issues.append(Issue(
                    IssueKind.FLAG_TYPO, flag, start + i,
                    f"Unknown flag '{flag}' — did you mean {near[0]}?",
                    tuple(near),
                ))
        i += 1

    if i >= len(rest):
        issues.append(Issue(
            IssueKind.ARG_REQUIRED, spec.name, start - 1,
            f"`{spec.name}` needs a package or binary to run",
            (),
        ))
        return issues

    pkg = rest[i]
    if looks_like_package(pkg):
        issues.append(Issue(
            IssueKind.PACKAGE_TYPO, pkg, start + i,
            f"Check package name '{pkg}'",
            (),
        ))
    return issues

File: go/detector.go
// Classification engine — port of src/fixpm/detector.py.
//
// One generic engine consumes every ManagerSpec; per-manager knowledge lives
// entirely in the generated rule tables. Kept line-for-line comparable with the
// Python source so a reviewer can diff the two side by side.
package main

import (
	"regexp"
	"sort"
	"strings"
)

// ---------------------------------------------------------------- model

// IssueKind mirrors fixpm.rules.base.IssueKind.
type IssueKind string

const (
	KindSubcommandTypo IssueKind = "subcommand-typo"
	KindFlagTypo       IssueKind = "flag-typo"
	KindMissingDashes  IssueKind = "missing-dashes"
	KindArgRequired    IssueKind = "argument-required"
	KindPackageTypo    IssueKind = "package-typo"
)

// KindLabel mirrors fixpm.rules.base.KIND_LABEL.
var KindLabel = map[IssueKind]string{
	KindSubcommandTypo: "command fix",
	KindFlagTypo:       "flag fix",
	KindMissingDashes:  "missing --",
	KindArgRequired:    "missing argument",
	KindPackageTypo:    "package fix",
}

// Issue mirrors fixpm.rules.base.Issue.
type Issue struct {
	Kind       IssueKind
	Token      string
	Index      int
	Message    string
	Candidates []string
}

// Correction mirrors fixpm.rules.base.Correction.
type Correction struct {
	Command    string
	Kind       IssueKind
	Score      float64
	Message    string
	Executable bool
}

// ---------------------------------------------------------------- detector

var pkgRe = regexp.MustCompile(
	`(?i)^(@[a-z0-9-~][a-z0-9-._~]*/)?[a-z0-9-~][a-z0-9-._~]*$`)

// LooksLikePackage mirrors detector.looks_like_package: a conservative
// npm-package-name shape check that rules out paths, globs and files.
func LooksLikePackage(token string) bool {
	if token == "" || strings.ContainsAny(token, "*~=\\\"") {
		return false
	}
	if strings.HasPrefix(token, ".") || strings.HasPrefix(token, "~") {
		return false
	}
	if strings.Contains(token, "/") && !strings.HasPrefix(token, "@") {
		return false
	}
	return pkgRe.MatchString(token)
}

// nearest mirrors detector._nearest. The pool is lowercased and sorted before
// ranking, exactly as the Python version does.
func nearest(token string, pool []string, top int, minScore float64) []string {
	lowered := make([]string, 0, len(pool))
	for _, p := range pool {
		lowered = append(lowered, strings.ToLower(p))
	}
	sort.Strings(lowered)
	return Rank(strings.ToLower(token), lowered, top, minScore)
}

// hinted mirrors detector._hinted: the curated hint for *token* if it is valid
// in this slot. One TypoHints map feeds both the command slot and the
// chain-verb slot, so the target is checked against the slot's own vocabulary.
func hinted(token string, spec *ManagerSpec, pool []string) []string {
	hint, ok := spec.TypoHints[strings.ToLower(token)]
	if !ok || !contains(pool, hint) {
		return nil
	}
	return []string{hint}
}

// subCandidates mirrors detector._sub_candidates.
func subCandidates(token string, spec *ManagerSpec) []string {
	if hint := hinted(token, spec, spec.Vocabulary()); hint != nil {
		return hint
	}
	return nearest(token, spec.Vocabulary(), 3, spec.SubcommandMinScore)
}

// Analyze mirrors detector.analyze.
func Analyze(text string, spec *ManagerSpec) ([]string, *ManagerSpec, []Issue) {
	tokens := Tokenize(text)
	index, located := locate(tokens, spec)
	if located == nil {
		return tokens, nil, nil
	}
	if index < 0 {
		// Forced table whose binary is absent: synthesise it so the fix is a
		// runnable command. See detector.analyze for the rationale.
		tokens = append([]string{located.Binaries[0]}, tokens...)
		index = 0
	}
	if located.PackageFirst {
		return tokens, located, analyzePackageFirst(located, tokens, index+1)
	}
	return tokens, located, analyzeSubcommand(located, tokens, index+1)
}

// locate mirrors detector._locate.
func locate(tokens []string, spec *ManagerSpec) (int, *ManagerSpec) {
	limit := len(tokens)
	if limit > 4 {
		limit = 4
	}
	if spec != nil {
		for i := 0; i < limit; i++ {
			for _, b := range spec.Binaries {
				if tokens[i] == b {
					return i, spec
				}
			}
		}
		// Binary absent: -1 tells Analyze to synthesise it. See
		// detector._locate for why returning nil here was wrong.
		return -1, spec
	}
	for i := 0; i < limit; i++ {
		if found := SpecForBinary(tokens[i]); found != nil {
			return i, found
		}
	}
	return -1, nil
}

// contains reports membership for the small string slices used by the rules.
func contains(list []string, want string) bool {
	for _, v := range list {
		if v == want {
			return true
		}
	}
	return false
}

// scanLeadingFlags mirrors the global-flag prelude shared by both analyzers.
func scanLeadingFlags(spec *ManagerSpec, rest []string, start int) ([]Issue, int) {
	issues := []Issue{}
	i := 0
	for i < len(rest) && strings.HasPrefix(rest[i], "-") {
		flag := rest[i]
		if contains(spec.ValueFlags, flag) && i+1 < len(rest) {
			i += 2
			continue
		}
		if !contains(spec.GlobalFlags, flag) {
			if near := nearest(flag, spec.GlobalFlags, 2, 0.6); len(near) > 0 {
				issues = append(issues, Issue{
					Kind:       KindFlagTypo,
					Token:      flag,
					Index:      start + i,
					Message:    "Unknown global flag '" + flag + "' — did you mean " + near[0] + "?",
					Candidates: near,
				})
			}
		}
		i++
	}
	return issues, i
}

// analyzeSubcommand mirrors detector._analyze_subcommand.
func analyzeSubcommand(spec *ManagerSpec, tokens []string, start int) []Issue {
	rest := tokens[start:]
	issues, i := scanLeadingFlags(spec, rest, start)

	if i >= len(rest) {
		return issues
	}

	sub := rest[i]
	subAbs := start + i
	canon, ok := spec.Canonical(sub)

	if !ok {
		if candidates := subCandidates(sub, spec); len(candidates) > 0 {
			issues = append(issues, Issue{
				Kind:       KindSubcommandTypo,
				Token:      sub,
				Index:      subAbs,
				Message:    "'" + sub + "' is not a " + spec.Name + " command",
				Candidates: candidates,
			})
		}
		// One fix at a time: stop validating after an unknown command.
		return issues
	}

	// Chain commands like `yarn global add <pkg>`, `docker container ls` or
	// `go mod tidy` — absorb the second-level verb. Which commands chain is a
	// per-CLI fact (spec.Chains); the slot gets typo tolerance too, so
	// `yarn global ad x` still resolves the chain and reports the typo.
	extra := 0
	verbs := spec.Chains[canon]
	if len(verbs) > 0 && i+1 < len(rest) {
		nxt := rest[i+1]
		if contains(verbs, nxt) {
			extra = 1
			canon = canon + " " + nxt
		} else {
			// Curated hints first (short transpositions such as `ud` -> `up`
			// fall under every floor), validated against this slot.
			near := hinted(nxt, spec, verbs)
			if near == nil {
				near = nearest(nxt, verbs, 1, spec.SubcommandMinScore)
			}
			if len(near) > 0 {
				issues = append(issues, Issue{
					Kind:       KindSubcommandTypo,
					Token:      nxt,
					Index:      start + i + 1,
					Message:    "'" + nxt + "' is not a " + spec.Name + " subcommand of '" + canon + "'",
					Candidates: near,
				})
				extra = 1
				canon = canon + " " + near[0]
			}
		}
	}

	effective := canon
	if idx := strings.LastIndex(canon, " "); idx >= 0 {
		effective = canon[idx+1:]
	}
	allowed := spec.Flags[effective]

	type positional struct {
		index int
		token string
	}
	positionals := []positional{}

	j := i + 1 + extra
	for j < len(rest) {
		token := rest[j]
		if token == "--" { // passthrough separator: everything after is exempt
			break
		}
		if strings.HasPrefix(token, "-") {
			if len(allowed) > 0 && !contains(allowed, token) &&
				!contains(spec.GlobalFlags, token) {
				if near := nearest(token, allowed, 2, 0.6); len(near) > 0 {
					issues = append(issues, Issue{
						Kind:  KindFlagTypo,
						Token: token,
						Index: start + j,
						Message: "Unknown flag '" + token + "' for `" + spec.Name +
							" " + effective + "` — did you mean " + near[0] + "?",
						Candidates: near,
					})
				}
			}
			if contains(spec.ValueFlags, token) {
				j++ // skip the consumed value
			}
		} else {
			positionals = append(positionals, positional{j, token})
		}
		j++
	}

	// `npm install express save-dev` -> missing dashes on `save-dev`
	dashless := make(map[string]string, len(allowed))
	for _, f := range allowed {
		dashless[strings.TrimLeft(f, "-")] = f
	}
	flagged := make(map[int]bool, len(positionals))
	for _, p := range positionals {
		if runeLen(p.token) > 1 {
			if want, hit := dashless[p.token]; hit {
				issues = append(issues, Issue{
					Kind:       KindMissingDashes,
					Token:      p.token,
					Index:      start + p.index,
					Message:    "Missing flag prefix — did you mean " + want + "?",
					Candidates: []string{want},
				})
				flagged[p.index] = true
			}
		}
	}
	kept := positionals[:0:0]
	for _, p := range positionals {
		if !flagged[p.index] {
			kept = append(kept, p)
		}
	}
	positionals = kept

	// Only report a missing argument when nothing else was flagged. For
	// `npm uninstall save-dev` the real problem is the missing dashes, and
	// pairing it with "needs an argument" emitted a second suggestion that
	// still contained the very typo the first one fixes.
	if contains(spec.ArgRequired, canon) && len(positionals) == 0 &&
		len(flagged) == 0 && extra == 0 {
		issues = append(issues, Issue{
			Kind:  KindArgRequired,
			Token: sub,
			Index: subAbs,
			Message: "`" + spec.Name + " " + sub + "` needs an argument (see `" +
				spec.Name + " help " + effective + "`)",
		})
	}

	// Chained package commands are listed explicitly ("global add"), so this
	// stays a plain membership test. See detector._analyze_subcommand.
	if contains(spec.PackageCommands, canon) {
		checked := 0
		for _, p := range positionals {
			if checked >= 2 {
				break
			}
			if contains(spec.Vocabulary(), strings.ToLower(p.token)) ||
				!LooksLikePackage(p.token) {
				continue
			}
			issues = append(issues, Issue{
				Kind:    KindPackageTypo,
				Token:   p.token,
				Index:   start + p.index,
				Message: "Check package name '" + p.token + "'",
			})
			checked++
		}
	}
	return issues
}

// analyzePackageFirst mirrors detector._analyze_package_first.
func analyzePackageFirst(spec *ManagerSpec, tokens []string, start int) []Issue {
	rest := tokens[start:]
	issues, i := scanLeadingFlags(spec, rest, start)

	if i >= len(rest) {
		issues = append(issues, Issue{
			Kind:  KindArgRequired,
			Token: spec.Name,
			Index: start - 1,
			Message: "`" + spec.Name +
				"` needs a package or binary to run",
		})
		return issues
	}

	pkg := rest[i]
	if LooksLikePackage(pkg) {
		issues = append(issues, Issue{
			Kind:    KindPackageTypo,
			Token:   pkg,
			Index:   start + i,
			Message: "Check package name '" + pkg + "'",
		})
	}
	return issues
}
